File: android/app/src/main/java/in/unsigned/keyboard/Transliterator.kt
package `in`.unsigned.keyboard

data class TransliterationResult(
    val transliterated: String,
    val consumed: Int,
    val remaining: String
)

class Transliterator {

    private val consonants = mapOf(
        "ক" to listOf("k", "K"),
        "খ" to listOf("kh", "kH", "Kh", "KH"),
        "গ" to listOf("g", "G"),
        "ঘ" to listOf("gh", "gH", "Gh", "GH"),
        "ঙ" to listOf("ng", "Ng", "NG", "nG"),
        "চ" to listOf("s", "c"),
        "ছ" to listOf("S", "ss", "Ss", "SS", "cc", "Cc", "CC"),
        "জ" to listOf("j", "z"),
        "ঝ" to listOf("jh", "jH", "Jh", "JH"),
        "ঞ" to listOf("yy", "Y"),
        "ট" to listOf("T"),
        "ঠ" to listOf("Th"),
        "ড" to listOf("D"),
        "ঢ" to listOf("Dh"),
        "ণ" to listOf("N"),
        "ত" to listOf("t"),
        "থ" to listOf("th"),
        "দ" to listOf("d"),
        "ধ" to listOf("dh"),
        "ন" to listOf("n"),
        "প" to listOf("p", "P"),
        "ফ" to listOf("f", "F", "ph", "pH", "Ph", "PH"),
        "ব" to listOf("b", "B"),
        "ভ" to listOf("v", "V", "bh", "bH", "Bh", "BH"),
        "ম" to listOf("m", "M"),
        "য" to listOf("J", "jj", "JJ", "Jj", "Z", "zz", "ZZ", "Zz"),
        "ৰ" to listOf("r"),
        "ল" to listOf("l", "L"),
        "ৱ" to listOf("w", "W"),
        "স" to listOf("x", "X"),
        "শ" to listOf("xx", "Xx", "xX", "XX"),
        "ষ" to listOf("xxx", "XxX", "xXX", "XXX", "xxX", "xXx", "Xxx"),
        "হ" to listOf("h", "H"),
        "ড়" to listOf("rr", "R", "RR", "rR", "Rr"),
        "ঢ়" to listOf("rh", "Rh", "rH", "RH"),
        "য়" to listOf("y"),
        "ক্ষ" to listOf("khyy")
    )

    private val vowels = mapOf(
        "অ" to listOf("o"),
        "আ" to listOf("a"),
        "ই" to listOf("i"),
        "ঈ" to listOf("ii"),
        "উ" to listOf("u"),
        "ঊ" to listOf("uu", "U"),
        "ঋ" to listOf("riii"),
        "এ" to listOf("e"),
        "ঐ" to listOf("oi"),
        "ও" to listOf("uuu", "oo", "O"),
        "ঔ" to listOf("ou")
    )

    private val vowelMarks = mapOf(
        "" to listOf("o"),
        "া" to listOf("a"),
        "ি" to listOf("i"),
        "ী" to listOf("ii"),
        "ু" to listOf("u"),
        "ূ" to listOf("uu", "U"),
        "ৃ" to listOf("riii"),
        "ে" to listOf("e"),
        "ৈ" to listOf("oi"),
        "ো" to listOf("uuu", "oo", "O"),
        "ৌ" to listOf("ou")
    )

    private val specialChar = mapOf(
        "ং" to listOf("ngg", "Ngg", "nGg", "NGg", "ngG", "NgG", "nGG", "NGG"),
        "ঃ" to listOf("hh", "HH", "hH", "Hh"),
        "ঁ" to listOf("*"),
        "ৎ" to listOf("t", "T"),
        "।" to listOf("|")
    )

    private val digits = mapOf(
        "০" to listOf("0"),
        "১" to listOf("1"),
        "২" to listOf("2"),
        "৩" to listOf("3"),
        "৪" to listOf("4"),
        "৫" to listOf("5"),
        "৬" to listOf("6"),
        "৭" to listOf("7"),
        "৮" to listOf("8"),
        "৯" to listOf("9")
    )

    private val specialCombinations = mapOf(
        "gyy" to "জ্ঞ",
        "jyy" to "হ্য",
        "khm" to "ক্ষ্ম"
    )

    private val explicitHolonto = "্"

    fun transliterate(input: String): String {
        var output = ""
        var i = 0
        var previousCharWasConsonant = false
        var skipNextCombination = false

        while (i < input.length) {
            val result = processNextCharacter(input, i, previousCharWasConsonant, skipNextCombination)
            output += result.output
            i = result.newIndex
            previousCharWasConsonant = result.previousCharWasConsonant
            skipNextCombination = result.skipNextCombination
        }

        return output
    }

    fun transliterateBuffer(input: String): TransliterationResult {
        val transliterated = transliterate(input)
        return TransliterationResult(transliterated, input.length, "")
    }

    private data class ProcessResult(
        val output: String,
        val newIndex: Int,
        val previousCharWasConsonant: Boolean,
        val skipNextCombination: Boolean
    )

    private fun processNextCharacter(
        input: String,
        index: Int,
        previousCharWasConsonant: Boolean,
        skipNextCombination: Boolean
    ): ProcessResult {

        if (input[index] == '.') {
            if (index + 1 < input.length && input[index + 1] == '.') {
                return ProcessResult(explicitHolonto, index + 2, false, false)
            }
            return ProcessResult("", index + 1, previousCharWasConsonant, true)
        }

        // Check special combinations
        for ((combo, result) in specialCombinations) {
            if (input.substring(index).startsWith(combo)) {
                return ProcessResult(result, index + combo.length, true, false)
            }
        }

        val longestMatch = findLongestMatch(input, index, skipNextCombination)

        if (longestMatch != null) {
            return processMatch(longestMatch.matchedChar, longestMatch.category,
                previousCharWasConsonant, skipNextCombination, longestMatch.match.length, index)
        }

        return ProcessResult(input[index].toString(), index + 1, false, false)
    }

    private data class MatchResult(
        val match: String,
        val matchedChar: String,
        val category: String
    )

    private fun findLongestMatch(input: String, index: Int, skipNextCombination: Boolean): MatchResult? {
        var longestMatch: MatchResult? = null
        val substring = input.substring(index)

        // Check consonants
        for ((assamese, romanizations) in consonants) {
            for (romanization in romanizations) {
                if (substring.startsWith(romanization) &&
                    (!skipNextCombination || romanization.length == 1) &&
                    (longestMatch == null || romanization.length > longestMatch.match.length)) {
                    longestMatch = MatchResult(romanization, assamese, "consonants")
                }
            }
        }

        // Check vowels
        for ((assamese, romanizations) in vowels) {
            for (romanization in romanizations) {
                if (substring.startsWith(romanization) &&
                    (!skipNextCombination || romanization.length == 1) &&
                    (longestMatch == null || romanization.length > longestMatch.match.length)) {
                    longestMatch = MatchResult(romanization, assamese, "vowels")
                }
            }
        }

        // Check special characters
        for ((assamese, romanizations) in specialChar) {
            for (romanization in romanizations) {
                if (substring.startsWith(romanization) &&
                    (!skipNextCombination || romanization.length == 1) &&
                    (longestMatch == null || romanization.length > longestMatch.match.length)) {
                    longestMatch = MatchResult(romanization, assamese, "specialChar")
                }
            }
        }

        // Check digits
        for ((assamese, romanizations) in digits) {
            for (romanization in romanizations) {
                if (substring.startsWith(romanization) &&
                    (!skipNextCombination || romanization.length == 1) &&
                    (longestMatch == null || romanization.length > longestMatch.match.length)) {
                    longestMatch = MatchResult(romanization, assamese, "digits")
                }
            }
        }

        return longestMatch
    }

    private fun processMatch(
        matchedChar: String,
        category: String,
        previousCharWasConsonant: Boolean,
        skipNextCombination: Boolean,
        matchLength: Int,
        index: Int
    ): ProcessResult {
        var output = ""
        var newPreviousCharWasConsonant = false

        // Debug logging
        println("DEBUG: matchedChar=$matchedChar, category=$category, previousCharWasConsonant=$previousCharWasConsonant")

        when (category) {
            "consonants" -> {
                if (previousCharWasConsonant && !skipNextCombination) {
                    output += explicitHolonto
                }
                output += matchedChar
                newPreviousCharWasConsonant = true
                println("DEBUG: consonant output=$output")
            }
            "vowels" -> {
                if (previousCharWasConsonant && !skipNextCombination) {
                    // We need to find the vowel mark for this vowel
                    // The matchedChar is the full vowel (like "ই"), we need the mark (like "ি")
                    val vowelMark = when (matchedChar) {
                        "আ" -> "া"
                        "ই" -> "ি"
                        "ঈ" -> "ী"
                        "উ" -> "ু"
                        "ঊ" -> "ূ"
                        "ঋ" -> "ৃ"
                        "এ" -> "ে"
                        "ঐ" -> "ৈ"
                        "ও" -> "ো"
                        "ঔ" -> "ৌ"
                        "অ" -> "" // No mark for 'o' sound
                        else -> matchedChar
                    }
                    output += vowelMark
                    println("DEBUG: vowel $matchedChar -> mark '$vowelMark', output=$output")
                } else {
                    output += matchedChar
                    println("DEBUG: standalone vowel output=$output")
                }
            }
            else -> {
                output += matchedChar
            }
        }

        return ProcessResult(output, index + matchLength, newPreviousCharWasConsonant, false)
    }
}

File: core/src/transliterator.rs
use crate::types::{CharCategory, MatchResult, TransliterationSchema};

/// Main transliterator struct
pub struct Transliterator {
    schema: TransliterationSchema,
}

impl Transliterator {
    /// Creates a new Transliterator with the given schema
    pub fn new(schema: TransliterationSchema) -> Self {
        Self { schema }
    }

    /// Transliterates input text according to the schema
    pub fn transliterate(&self, input: &str) -> String {
        let mut output = String::new();
        let mut i = 0;
        let chars: Vec<char> = input.chars().collect();
        let mut previous_char_was_consonant = false;
        let mut skip_next_combination = false;
        let mut in_backticks = false;

        while i < chars.len() {
            let result = self.process_next_character(
                &chars,
                i,
                previous_char_was_consonant,
                skip_next_combination,
                in_backticks,
            );

            output.push_str(&result.output);
            i = result.new_index;
            previous_char_was_consonant = result.previous_char_was_consonant;
            skip_next_combination = result.skip_next_combination;
            in_backticks = result.in_backticks;
        }

        output
    }

    /// Gets a reference to the current schema
    pub fn get_schema(&self) -> &TransliterationSchema {
        &self.schema
    }

    /// Sets a new schema
    pub fn set_schema(&mut self, schema: TransliterationSchema) {
        self.schema = schema;
    }

    fn process_next_character(
        &self,
        chars: &[char],
        index: usize,
        previous_char_was_consonant: bool,
        skip_next_combination: bool,
        in_backticks: bool,
    ) -> ProcessResult {
        // Handle backticks for escaping
        if let Some(result) = self.handle_backticks(chars, index, in_backticks) {
            return result;
        }

        // If inside backticks, pass through character as-is
        if in_backticks {
            return ProcessResult {
                output: chars[index].to_string(),
                new_index: index + 1,
                previous_char_was_consonant,
                skip_next_combination,
                in_backticks,
            };
        }

        // Handle explicit halant (..)
        if chars[index] == '.' {
            if index + 1 < chars.len() && chars[index + 1] == '.' {
                return ProcessResult {
                    output: self.schema.exceptions.explicit_holonto.clone(),
                    new_index: index + 2,
                    previous_char_was_consonant: false,
                    skip_next_combination: false,
                    in_backticks,
                };
            }
            // Single dot - skip next combination
            return ProcessResult {
                output: String::new(),
                new_index: index + 1,
                previous_char_was_consonant,
                skip_next_combination: true,
                in_backticks,
            };
        }

        // Check for special combinations
        let remaining: String = chars[index..].iter().collect();
        for (combo, result) in &self.schema.exceptions.special_combinations {
            if remaining.starts_with(combo) {
                return ProcessResult {
                    output: result.clone(),
                    new_index: index + combo.chars().count(),
                    previous_char_was_consonant: true,
                    skip_next_combination: false,
                    in_backticks: false,
                };
            }
        }

        // Handle regular transliteration
        self.handle_regular_transliteration(
            chars,
            index,
            previous_char_was_consonant,
            skip_next_combination,
        )
    }

    fn handle_backticks(
        &self,
        chars: &[char],
        index: usize,
        in_backticks: bool,
    ) -> Option<ProcessResult> {
        // Handle escaped backtick \`
        if chars[index] == '\\' && index + 1 < chars.len() && chars[index + 1] == '`' {
            return Some(ProcessResult {
                output: "`".to_string(),
                new_index: index + 2,
                previous_char_was_consonant: false,
                skip_next_combination: false,
                in_backticks,
            });
        }

        // Toggle backtick mode
        if chars[index] == '`' {
            return Some(ProcessResult {
                output: String::new(),
                new_index: index + 1,
                previous_char_was_consonant: false,
                skip_next_combination: false,
                in_backticks: !in_backticks,
            });
        }

        None
    }

    fn handle_regular_transliteration(
        &self,
        chars: &[char],
        index: usize,
        previous_char_was_consonant: bool,
        skip_next_combination: bool,
    ) -> ProcessResult {
        let remaining: String = chars[index..].iter().collect();

        // Find the longest match
        if let Some(match_result) = self.find_longest_match(&remaining, skip_next_combination) {
            return self.process_match(
                chars,
                index,
                &match_result,
                previous_char_was_consonant,
                skip_next_combination,
            );
        }

        // No match found - pass through the character
        ProcessResult {
            output: chars[index].to_string(),
            new_index: index + 1,
            previous_char_was_consonant: false,
            skip_next_combination: false,
            in_backticks: false,
        }
    }

    fn find_longest_match(&self, input: &str, skip_next_combination: bool) -> Option<MatchResult> {
        let mut longest_match: Option<MatchResult> = None;
        let mut max_length = 0;

        // Helper function to check a category
        let check_category = |map: &std::collections::HashMap<String, Vec<String>>,
                              category: CharCategory,
                              current_max: usize| -> Option<MatchResult> {
            let mut best: Option<MatchResult> = None;
            let mut best_len = current_max;

            for (target_char, romanizations) in map {
                for romanization in romanizations {
                    if input.starts_with(romanization) {
                        let len = romanization.len();
                        if len > best_len && (!skip_next_combination || len == 1) {
                            best = Some(MatchResult {
                                matched_char: target_char.clone(),
                                matched_length: len,
                                category,
                            });
                            best_len = len;
                        }
                    }
                }
            }
            best
        };

        // Check all categories
        let categories = [
            (&self.schema.consonants, CharCategory::Consonant),
            (&self.schema.vowels, CharCategory::Vowel),
            (&self.schema.vowel_marks, CharCategory::VowelMark),
            (&self.schema.special_char, CharCategory::SpecialChar),
            (&self.schema.digits, CharCategory::Digit),
        ];

        for (map, category) in categories {
            if let Some(result) = check_category(map, category, max_length) {
                if result.matched_length > max_length {
                    max_length = result.matched_length;
                    longest_match = Some(result);
                }
            }
        }

        longest_match
    }

    fn process_match(
        &self,
        chars: &[char],
        index: usize,
        match_result: &MatchResult,
        previous_char_was_consonant: bool,
        skip_next_combination: bool,
    ) -> ProcessResult {
        let mut output = String::new();
        let mut new_previous_char_was_consonant = false;

        match match_result.category {
            CharCategory::Consonant => {
                // Add explicit halant if previous was consonant
                if previous_char_was_consonant && !skip_next_combination {
                    output.push_str(&self.schema.exceptions.explicit_holonto);
                }

                // Check if next character is a consonant for joined forms
                let next_consonant = self.find_next_consonant(chars, index + match_result.matched_length);

                if next_consonant.is_some()
                    && self.schema.exceptions.joined_consonants_before.contains_key(&match_result.matched_char) {
                    output.push_str(&self.schema.exceptions.joined_consonants_before[&match_result.matched_char]);
                } else {
                    output.push_str(&match_result.matched_char);
                }

                new_previous_char_was_consonant = true;
            }
            CharCategory::Vowel => {
                if previous_char_was_consonant && !skip_next_combination {
                    // Find corresponding vowel mark
                    let mut found_mark = false;
                    for (vowel_mark, romanizations) in &self.schema.vowel_marks {
                        if let Some(first_rom) = self.schema.vowels.get(&match_result.matched_char).and_then(|v| v.first()) {
                            if romanizations.contains(first_rom) {
                                output.push_str(vowel_mark);
                                found_mark = true;
                                break;
                            }
                        }
                    }
                    if !found_mark {
                        output.push_str(&match_result.matched_char);
                    }
                } else {
                    output.push_str(&match_result.matched_char);
                }
            }
            _ => {
                output.push_str(&match_result.matched_char);
            }
        }

        // Check for joined consonants after
        let romanization: String = chars[index..index + match_result.matched_length].iter().collect();
        if previous_char_was_consonant
            && self.schema.exceptions.joined_consonants_after.contains_key(&romanization) {
            output = format!(
                "{}{}",
                self.schema.exceptions.explicit_holonto,
                self.schema.exceptions.joined_consonants_after[&romanization]
            );
            new_previous_char_was_consonant = true;
        }

        ProcessResult {
            output,
            new_index: index + match_result.matched_length,
            previous_char_was_consonant: new_previous_char_was_consonant,
            skip_next_combination: false,
            in_backticks: false,
        }
    }

    fn find_next_consonant(&self, chars: &[char], start_index: usize) -> Option<String> {
        for i in start_index..chars.len() {
            let remaining: String = chars[i..].iter().collect();
            if let Some(match_result) = self.find_longest_match(&remaining, false) {
                if match_result.category == CharCategory::Consonant {
                    return Some(match_result.matched_char);
                }
                // If we matched something that's not a consonant, stop searching
                break;
            }
        }
        None
    }
}

/// Result of processing a character
struct ProcessResult {
    output: String,
    new_index: usize,
    previous_char_was_consonant: bool,
    skip_next_combination: bool,
    in_backticks: bool,
}

/// Standalone function to transliterate with a schema
pub fn transliterate(input: &str, schema: &TransliterationSchema) -> String {
    let t = Transliterator::new(schema.clone());
    t.transliterate(input)
}
